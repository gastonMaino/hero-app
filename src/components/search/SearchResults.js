import React, { useMemo } from 'react';
import { HeroeContext } from '../heroes/HeroContext';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import PropTypes from 'prop-types';

export const SearchResults = ({q}) => {

    const heroData = useMemo(() => getHeroesByName(q), [q]);


    return (
        <section className='section-results'>
            <h2>Results: </h2>

                {
                    (q === '')
                    &&
                    <div className='alert'>
                        <p className='alert__paragraph'>
                            Search a hero!
                        </p>
                    </div>
                }

                {
                    (q !== '' && heroData.length === 0)
                    &&
                    <div className='alert-danger'>
                        <p className='alert-danger__paragraph'>
                            There is no a hero {q}
                        </p>
                    </div>
                }

                {
                    heroData.map((hero) => {
                        return (
                            <HeroeContext.Provider
                                key={hero.id} value={hero}
                            >
                                <HeroCard />
                            </HeroeContext.Provider>
                        )

                    })
                }    
        </section>
    )
}

SearchResults.propTypes = {
    q: PropTypes.string.isRequired
}
