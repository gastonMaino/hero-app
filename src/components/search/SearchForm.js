import React from 'react';
import queryString from 'query-string'
import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm/useForm'
import { SearchResults } from './SearchResults';


export const SearchForm = ({ history }) => {


    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);

    const [{ search }, handleInputChange] = useForm({
        search: q
    })

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${search}`)
    }

    return (
        <>
            <section className='search-form'>
                <h2 className='search-form__title'>Search Form:</h2>
                <form onSubmit={handleSearch} className='form'>
                    <input
                        type='text'
                        name='search'
                        value={search}
                        placeholder='find your hero'
                        aria-label='find your hero'
                        className='form__input'
                        autoComplete='off'
                        onChange={handleInputChange}
                    />
                    <button className='btn-search'>
                        Search
                    </button>
                </form>
            </section>

            <SearchResults q={q} />
        </>
    )
}
