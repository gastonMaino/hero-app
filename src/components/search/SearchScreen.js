import React from 'react';
import { SearchForm } from './SearchForm';

export const SearchScreen = ({ history }) => {
    return (
        <article className='search'>
            <div className='container-search-title'>
                <h1 className='search-title'>Search</h1>
            </div>

            <div className='container-items-search'>
                <SearchForm history={history} />
            </div>
        </article>
    )
}
