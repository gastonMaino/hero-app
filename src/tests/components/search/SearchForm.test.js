import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { SearchForm } from "../../../components/search/SearchForm"

describe('test in <SearchForm />', () => {


    test('should match to snapshot', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path='/search' component={SearchForm} />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert__paragraph').text().trim()).toBe('Search a hero!');
    })

    test('should show batman and input value have the querySrting', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path='/search' component={SearchForm} />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper.find('HeroCard').exists()).toBe(true);
    })

    test('should show message error', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman1233']}>
                <Route path='/search' component={SearchForm} />
            </MemoryRouter>
        )

        expect(wrapper.find('.alert-danger__paragraph').text().trim()).toBe('There is no a hero batman1233');
    })

    test('should call history.push', () => {

        const historyMock = {
            push: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route
                    path='/search'
                    component={() => <SearchForm history={historyMock} />}
                />
            </MemoryRouter>
        )

        wrapper.find('input').simulate('change', {
            target: {
                name: 'search',
                value: 'batman'
            }
        });
 
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
 
        expect( historyMock.push ).toHaveBeenCalledWith(`?q=batman`);

    })



})
