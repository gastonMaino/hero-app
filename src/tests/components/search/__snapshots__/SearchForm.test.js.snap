// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <SearchForm /> should match to snapshot 1`] = `
Array [
  <section>
    <h2>
      Search Form:
    </h2>
    <form
      className="form"
      onSubmit={[Function]}
    >
      <input
        aria-label="find your hero"
        autoComplete="off"
        className="form__input"
        name="search"
        onChange={[Function]}
        placeholder="find your hero"
        type="text"
        value=""
      />
      <button
        className="btn"
      >
        Search
      </button>
    </form>
  </section>,
  <section
    className="section-results"
  >
    <h2>
      Results: 
    </h2>
    <div
      className="alert"
    >
      <p
        className="alert__paragraph"
      >
        Search a hero!
      </p>
    </div>
  </section>,
]
`;

exports[`test in <SearchForm /> should show batman and input value have the querySrting 1`] = `
Array [
  <section>
    <h2>
      Search Form:
    </h2>
    <form
      className="form"
      onSubmit={[Function]}
    >
      <input
        aria-label="find your hero"
        autoComplete="off"
        className="form__input"
        name="search"
        onChange={[Function]}
        placeholder="find your hero"
        type="text"
        value="batman"
      />
      <button
        className="btn"
      >
        Search
      </button>
    </form>
  </section>,
  <section
    className="section-results"
  >
    <h2>
      Results: 
    </h2>
    <div
      className="card"
    >
      <div
        className="card-image"
      >
        <img
          alt="Is a hero Batman"
          src="assets/heroes/dc-batman.jpg"
        />
      </div>
      <div
        className="card-content"
      >
        Array [
          <p
            className="title"
          >
            Batman
          </p>,
          <p
            className="paragraph"
          >
            Bruce Wayne
          </p>,
          <p
            className="paragraph"
          >
            <span
              className="paragraph__span"
            >
              Detective Comics #27
            </span>
          </p>,
          <a
            className="button"
            href="/hero/dc-batman"
            onClick={[Function]}
          >
            See more...
          </a>,
        ]
      </div>
    </div>
  </section>,
]
`;
